homie-controller/src/lib.rs

```diff
@@ -170,7 +169,34 @@ impl HomieController {
         Ok(None)
     }
 
+    /// Handle a publish event received from the MQTT broker, updating the devices and our
+    /// subscriptions as appropriate and possibly returning an event to send back to the controller
+    /// application.
     async fn handle_publish(&self, publish: Publish) -> Result<Option<Event>, HandleError> {
+        let (event, topics_to_subscribe, topics_to_unsubscribe) =
+            self.handle_publish_update_devices(publish)?;
+
+        for topic in topics_to_subscribe {
+            log::trace!("Subscribe to {}", topic);
+            self.mqtt_client.subscribe(topic, QoS::AtLeastOnce).await?;
+        }
+        for topic in topics_to_unsubscribe {
+            log::trace!("Unsubscribe from {}", topic);
+            self.mqtt_client.unsubscribe(topic).await?;
+        }
+
+        Ok(event)
+    }
+
+    /// Handle a publish event, update the devices, and return any event and any new topics which
+    /// should be subscribed to or unsubscribed from.
+    ///
+    /// This is separate from `handle_publish` because it takes the `devices` lock, to ensure that
+    /// no async operations are awaited while the lock is held.
+    fn handle_publish_update_devices(
```

Comment:
This method needs a better name.

Reply:
I would be okay with handle_publish_sync or update_devices. It's a bit fishy that it mutates things and also returns something. I suppose it could receive devices by ownership and return devices by ownership, along with the other arguments and return types, and then the name would be less important, because the arguments would tell you what it's doing.

Reply:
Yeah, the split is really just because it does the parts of `handle_publish` which don't need to call `.await`. It could be a block within `handle_publish`, except that it would be too easy for someone to later accidentally add a `.await` in where it shouldn't be.